src/proto/network/snp.rs

```diff
@@ -0,0 +1,383 @@
+//! Network I/O protocols.
+
+use crate::proto::Protocol;
+use crate::{unsafe_guid, Result, Status};
+
+/// The Snp protocol.
+#[repr(C)]
+#[unsafe_guid("a19832b9-ac25-11d3-9a2d-0090273fc14d")]
+#[derive(Protocol)]
+pub struct Snp {
+    pub revision: u64,
+    start: extern "efiapi" fn(this: &Snp) -> Status,
+    stop: extern "efiapi" fn(this: &Snp) -> Status,
+    initialize: extern "efiapi" fn(
+        this: &Snp,
+        extra_rx_buffer_size: usize,
+        extra_tx_buffer_size: usize,
+    ) -> Status,
+    reset: extern "efiapi" fn(this: &Snp, extended_verification: bool) -> Status,
+    shutdown: extern "efiapi" fn(this: &Snp) -> Status,
+    receive_filters: extern "efiapi" fn(
+        this: &Snp,
+        enable: u32,
+        disable: u32,
+        reset_mcast_filter: bool,
+        mcast_filter_count: usize,
+        mcast_filter: *const MacAddress,
+    ) -> Status,
+    station_address: extern "efiapi" fn(this: &Snp, reset: bool, new: *const MacAddress) -> Status,
+    statistics: extern "efiapi" fn(
+        this: &Snp,
+        reset: bool,
+        statistics_size: *mut usize,
+        statistics_table: *mut NetworkStatistics,
+    ) -> Status,
+    mcast_ip_to_mac: extern "efiapi" fn(this: &Snp, ipv6: bool, ip: *const IpAddress, mac: *mut MacAddress) -> Status,
+    nv_data: extern "efiapi" fn(
+        this: &Snp,
+        read_write: bool,
+        offset: usize,
+        buffer_size: usize,
+        buffer: *mut [u8],
+    ) -> Status,
+    get_status:
+        extern "efiapi" fn(this: &Snp, interrupt_status: *mut u32, tx_buf: *mut *mut [u8]) -> Status,
+    transmit: extern "efiapi" fn(
+        this: &Snp,
+        header_size: usize,
+        buffer_size: usize,
+        buffer: *const [u8],
+        src_addr: *const MacAddress,
+        dest_addr: *const MacAddress,
+        protocol: u16,
+    ) -> Status,
+    receive: extern "efiapi" fn(
+        this: &Snp,
+        header_size: *const usize,
+        buffer_size: *mut usize,
+        buffer: *mut [u8],
+        src_addr: *mut MacAddress,
+        dest_addr: *mut MacAddress,
+        protocol: *mut u16,
+    ) -> Status,
+    wait_for_packet: usize,
+    mode: *const NetworkMode,
+}
+
+impl Snp {
+    /// Changes the state of a network interface from “stopped” to “started”.
+    ///
+    /// # Errors
+    /// * `uefi::Status::ALREADY_STARTED`  The network interface is already in the started state.
+    /// * `uefi::Status::INVALID_PARAMETER`  This parameter was NULL or did not point to a valid EFI_SIMPLE_NETWORK_PROTOCOL structure
```

Comment:
Since we're assuming this method is called from safe Rust with a valid protocol object, is this error case possible?